### ISSUE TITLE
[JVM] NullPointerException when handling exception from unsupported revision

### ISSUE THREAD

**Issue by usev6** [rakudo/rakudo#5180]
## The Problem
Nowadays the error reporting from trying to use an unsupported revision leads to a NullPointerException on the JVM backend.
The following example is taken from one of Rakudo's tests -- which now fails on the JVM backend: https://github.com/rakudo/rakudo/blob/c88f50d5e8/t/02-rakudo/14-revisions.t#L32-L36.
```
$ ./rakudo-j -e 'use v6.c.NOMOD'
Unhandled exception: java.lang.NullPointerException
  in DYNAMIC (gen/jvm/CORE.c.setting:849)
  in print_exception (gen/jvm/CORE.c.setting:53658)
  in <anon> (gen/jvm/CORE.c.setting:53758)
  in command_eval (gen/jvm/stage2/NQPHLL.nqp:1394)
  in command_eval (gen/jvm/Compiler.nqp:236)
  in command_line (gen/jvm/stage2/NQPHLL.nqp:1315)
  in MAIN (gen/jvm/rakudo.nqp:151)
  in <mainline> (gen/jvm/rakudo.nqp:130)
  in <anon> (gen/jvm/rakudo.nqp)
```

This used to work in the past, but I didn't try to figure out when it broke.

Interestingly, there is no internal error when using ```--ll-exception```:
```
$ ./rakudo-j --ll-exception -e 'use v6.c.NOMOD'
No compiler available for Raku v6.c.NOMOD
  in throw (gen/jvm/CORE.c.setting:53196)
  in throw (gen/jvm/World.nqp:5483)
  in typed_panic (gen/jvm/Grammar.nqp:296)
  in !check-version-modifier (gen/jvm/World.nqp:557)
  in load-lang-ver (gen/jvm/World.nqp:596)
  in lang-version (gen/jvm/Grammar.nqp:872)
  in comp_unit (gen/jvm/Grammar.nqp:838)
  in TOP (gen/jvm/Grammar.nqp:557)
  in parse (gen/jvm/stage2/QRegex.nqp:2270)
  in parse (gen/jvm/stage2/NQPHLL.nqp:1579)
  in execute_stage (gen/jvm/stage2/NQPHLL.nqp:1491)
  in run (gen/jvm/stage2/NQPHLL.nqp:1527)
  in <anon> (gen/jvm/stage2/NQPHLL.nqp:1530)
  in compile (gen/jvm/stage2/NQPHLL.nqp:1522)
  in eval (gen/jvm/stage2/NQPHLL.nqp:1192)
  in <anon> (gen/jvm/stage2/NQPHLL.nqp:1335)
  in command_eval (gen/jvm/stage2/NQPHLL.nqp:1332)
  in command_eval (gen/jvm/Compiler.nqp:236)
  in command_line (gen/jvm/stage2/NQPHLL.nqp:1315)
  in MAIN (gen/jvm/rakudo.nqp:151)
  in <mainline> (gen/jvm/rakudo.nqp:130)
  in <anon> (gen/jvm/rakudo.nqp)
```

I'm opening this issue to document the problem. I plan to look closer, once other problems (like the broken build, cmp. https://github.com/rakudo/rakudo/pull/5164) are fixed.

A good entry point for debugging could be https://github.com/Raku/nqp/blob/081af2885f/src/HLL/Compiler.nqp#L361-L375.

## Environment
* Operating system: FreeBSD 12.4-RELEASE
* Compiler version (`perl6 -v` or `raku -v`): Welcome to Rakudo™ v2022.12-1084-g7702e6b7d.
Implementing the Raku® Programming Language v6.d.
Built on JVM.

**Comment by usev6** [comment]
Some observations:

For some reasons ```GLOBAL``` seems to be ```null``` here: https://github.com/rakudo/rakudo/blob/37f5259958/src/core.c/stubs.pm6#L53

Sidestepping that problem there is another problem with ```$*ERR```: The call to ```$err.say``` in https://github.com/rakudo/rakudo/blob/37f5259958/src/core.c/Exception.pm6#L555 fails. (It complains about "Unhandled exception; category = 1".) Replacing ```$err.say``` with ```nqp::say``` gives the correct result (but on stdout).

**Comment by usev6** [comment]
I've found another test that fails in the same way: https://github.com/Raku/roast/blob/54762714e5/integration/error-reporting.t#L200-L203
```
$ ./rakudo-j -e '#`{{ my long
    unfinished comment'
Unhandled exception: java.lang.NullPointerException
  in DYNAMIC (gen/jvm/CORE.c.setting:849)
  in print_exception (gen/jvm/CORE.c.setting:53657)
  in <anon> (gen/jvm/CORE.c.setting:53764)
  in command_eval (gen/jvm/stage2/NQPHLL.nqp:1394)
  in command_eval (gen/jvm/Compiler.nqp:236)
  in command_line (gen/jvm/stage2/NQPHLL.nqp:1315)
  in MAIN (gen/jvm/rakudo.nqp:151)
  in <mainline> (gen/jvm/rakudo.nqp:130)
  in <anon> (gen/jvm/rakudo.nqp)
```

**Comment by usev6** [comment]
I kind of rediscovered this bug today. If I'm not mistaken, ```GLOBAL``` is ```null``` at this place on MoarVM as well: https://github.com/rakudo/rakudo/blob/37f5259958/src/core.c/stubs.pm6#L53

But on MoarVM this doesn't blow up:
```
$ ./rakudo-m -e 'use nqp; my $foo := nqp::null; say $foo.WHO'
(Mu)
```
versus
```
$ ./rakudo-j -e 'use nqp; my $foo := nqp::null; dd $foo.WHO'
java.lang.NullPointerException
  in block <unit> at -e line 1
```

Now I wonder if it wouldn't make sense to add an additional check for ```GLOBAL``` being ```null```, e.g.:
```
diff --git a/src/core.c/stubs.rakumod b/src/core.c/stubs.rakumod
index d812c6ca8..93f0114d7 100644
--- a/src/core.c/stubs.rakumod
+++ b/src/core.c/stubs.rakumod
@@ -49,12 +49,13 @@ sub DYNAMIC(str $name, @deprecation?) is raw {  # is implementation-detail
           value,
           nqp::stmts(
             (my str $pkgname = nqp::replace($name,1,1,'')),
-            nqp::ifnull(
-              nqp::atkey(GLOBAL.WHO,$pkgname),
+            nqp::if(
+              nqp::isnull(GLOBAL) || nqp::isnull(my \res := nqp::atkey(GLOBAL.WHO,$pkgname)),
               nqp::ifnull(
                 nqp::atkey(PROCESS.WHO,$pkgname),
                 Rakudo::Internals.INITIALIZE-DYNAMIC($name, @deprecation)
-              )
+              ),
+              res
             )
           )
         )
```

**Comment by usev6** [comment]
I've created a PR with the above patch: https://github.com/rakudo/rakudo/pull/5379

Maybe it's not the right solution for the problem, but at least we could discuss this idea.

**Comment by usev6** [comment]
Fixed with e7f9223c59.

Failing spectest unfudged with https://github.com/Raku/roast/commit/0905a199d0
